### DISCUSSION TITLE
Protecting API routes allowing only calls inside my APP

### DISCUSSION THREAD

**Discussion by puntope** [vercel/next.js#19066] [Help] [answered]
Hello all!

I have this kind of file in my API directory, In which I do a post request to DynamoDb or other services I need.  
I save ofc the API_keys in env.local and so on... to hide them in the browser...

But if someone runs a POST request directly to /api/post in an external way the code is also running.

I checked the docs but I cannot find any way to protect my route allowing only calls inside my own APP/Website. I tried to check the origin but I'm getting my own server actually as origin. 

Is any possible solution to protect the route? 

api/post.js 

```
export default async (req, res) => {

    if (req.method !== 'POST') return false

    // I DONT WANT THIS TO BE EXECUTED EXCEPT IN MY OWN APP

    const { body } = req

    const endpoint = process.env.MY_ENDPOINT
    const apiKey = process.env.MY_APIKEY

    const params = await JSON.parse(body)

    const headers = {
        'Content-Type': 'application/json',
    }
    const requestUrl = `${endpoint}/integration?apiKey=apiKey`

    const apiRes = await fetch(requestUrl, {
        headers: headers,
        method: 'POST',
        body: JSON.stringify(params?.formBody),
    })

[....]
```

pages/page.js 

```
const Page = () => {
    [...]

    useEffect(() => {
        await fetch('/api/post', { ... })
   }, [submitted])  

   [...]

}
```

**Comment by jamesmosier** [comment]
Hi @puntope. API routes are [same-origin only](https://nextjs.org/docs/api-routes/introduction#caveats) by default. 

Although you can also setup [CORS support](https://nextjs.org/docs/api-routes/api-middlewares#connectexpress-middleware-support). There is an [example](https://github.com/vercel/next.js/blob/canary/examples/api-routes-cors/pages/api/cors.js) that shows this.

**Reply by jamesmosier** [reply]
@puntope see this answer: https://github.com/vercel/next.js/issues/17814#issuecomment-707029253

> CORS is a client-implemented protocol (e.g. Browsers). It does not block HTTP requests if you use curl / postman etc because those do not implement CORS.

**Reply by puntope** [reply]
Hey, thanks for the info. Didn't know that was any difference between those.

**Comment by puntope** [comment, marked as answer]
@jamesmosier Thanks for the answer. Unfortunately was not really helpful. As expressed above I already read the docs and the examples but was not very clear and that's why I ask here...  

If I run the POST request in POSTMAN for example. It runs. So I guess the API route is not really protected.

What other frameworks like Laravel do is to add a CSRF token in the client-side making it impossible to perform a fake POST request against the API. 

https://portswigger.net/web-security/csrf/tokens

**Reply by rafde** [reply]
@puntope that implementation is up to you. There are libraries like next-auth that help you with that sort of implementation.

**Reply by puntope** [reply]
Thanks, man, this helps :) Looks promising the library

**Reply by koji98** [reply]
@puntope Were you able to figure out a solution using next-auth? I've been struggling with this for a few hours.

**Comment by tilxyz** [comment]
> API Routes do not specify CORS headers, meaning they are same-origin only by default.

I'm not sure if the document [here](https://nextjs.org/docs/pages/building-your-application/routing/api-routes#caveats) is accurate.  I am able to make an API from a chrome extension in the browser, the origin is `chrome-extension://<chrome_extension_id>`, which is obviously different from my domain, but it still passes. Do we know why it's possible?

**Comment by silviuio** [comment]
Does anyone have an answer to this? I am still confused as to what is the way to protect my nextjs api routes from beign accessed from another PC, via POSTMAN or something similar. The official docs say nothing about this

**Reply by Namnp1521** [reply]
i got the same problem, need some advice pls

**Reply by luiscaceresd** [reply]
Same here
